bin/shpec

```diff
@@ -44,97 +54,138 @@ stub_command() {
-
+    print_result "case '$2' in *$3*) false ;; *) :;; esac" \
+      "Expected [$2] not to match [$3]"
+  ;;
   xpresent )
-    print_result "[[ -n '$2' ]]" "Expected [$2] to be present";;
-
+    print_result "[ -n '$2' ]" \
+      "Expected [$2] to be present"
+  ;;
   xblank )
-    print_result "[[ -z '$2' ]]" "Expected [$2] to be blank";;
+    print_result "[ -z '$2' ]" \
+      "Expected [$2] to be blank"
+  ;;
 
   xfile_present )
-    print_result "[[ -e $2 ]]" "Expected file [$2] to exist";;
-
+    print_result "[ -e $2 ]" \
+      "Expected file [$2] to exist"
+  ;;
   xfile_absent )
-    print_result "[[ ! -e $2 ]]" "Expected file [$2] not to exist";;
-
+    print_result "[ ! -e $2 ]" \
+      "Expected file [$2] not to exist"
+  ;;
   xsymlink )
     link="$(readlink $2)"
-    print_result "[[ '$link' = '$3' ]]" "Expected [$2] to link to [$3], but got [$link]";;
-
+    print_result "[ '$link' = '$3' ]" \
+      "Expected [$2] to link to [$3], but got [$link]"
+  ;;
   xtest )
-    print_result "$2" "Expected $2 to be true";;
-
+    print_result "$2" \
+      "Expected $2 to be true"
+  ;;
   * )
     if type "$1" 2>/dev/null | grep -q 'function'; then
       matcher="$1"; shift
       $matcher "$@"
     else
       print_result false "Error: Unknown matcher [$1]"
-    fi;;
-
+    fi
+  ;;
   esac
 }
 
 print_result() {
   if eval "$1"; then
-    iecho -e "$green$assertion$norm"
+    iecho "$green$assertion$norm"
   else
-    ((failures += 1))
-    iecho -e "$red$assertion"
-    iecho -e "($2)$norm"
+    : $((failures += 1))
+    iecho "$red$assertion"
+    iecho "($2)$norm"
   fi
 }
 
-SHPEC_ROOT=${SHPEC_ROOT:-$([[ -d './shpec' ]] && echo './shpec' || echo '.')}
-
-case "$1" in
-
-  -v|--version )
-
-    echo "$VERSION";;
-
-  * )
-
-    matcher_files="$(find "$SHPEC_ROOT/matchers" -name '*.sh' 2>/dev/null)"
-
-    for matcher_file in $matcher_files; do
-      . "$matcher_file"
-    done
-
-    files="${@:-$(find $SHPEC_ROOT -name '*_shpec.sh')}"
-
-    time for file in $files; do
-      . "$file"
-    done
 
-    [ $failures -eq 0 ] && color=$green || color=$red
-    echo -e "${color}${examples} examples, ${failures} failures${norm}"
+shpec() {
+  (
+  VERSION=0.1.2posix
+  examples=0
+  failures=0
+  test_indent=0
+  red="\033[0;31m"
+  green="\033[0;32m"
+  norm="\033[0m"
+
+  SHPEC_ROOT=${SHPEC_ROOT:-$([ -d './shpec' ] && echo './shpec' || echo '.')}
+  case "$1" in
+  ( -v | --version ) echo "$VERSION"
+  ;;
+  ( * )
+      matcher_files="$(
+          find "$SHPEC_ROOT/matchers" -name '*.sh' 2>/dev/null
+      )"
+
+      for matcher_file in $matcher_files; do
+        . "$matcher_file"
+      done
+
+      if [ $# -gt 0 ] ; then
+        : \$1="'$1'" \${#1}="'${#1}'"
+        files="${@}"
+      else
+        files=$(
+          find $SHPEC_ROOT -name '*_shpec.sh'
+        )
+      fi
+
+      for file in $files; do
+        echo >& 2 "$file"
+        . "$file" || exit
```

Comment:
On 04/13/15 04:07, Ryland Herrick wrote:

>> +
>> +      for file in $files; do
>> +        echo >& 2 "$file"
>> +        . "$file" || exit
>
> Can you explain the intent behind the ||| exit|?

Hmm... The idea was catching errors in sourcing $file.
The kind of errors where it doesn't make sense to
continue - serious syntax trouble, etc.

However, this would also consider any non-zero $failure count
an error, and stop processing.

I will think about alternatives here, but for now it just
has to go.

Cheers,
HL